### DISCUSSION TITLE
Constraining LLM output using a Lark grammar

### DISCUSSION THREAD

**Discussion by nchammas** [lark-parser/lark#1515] [General]
I am just starting to experiment with LLMs for the first time and need to constrain the model's output using a Lark grammar.

I noticed that @erezsh recently starred [parserLLM][1], but that project seems unmaintained.

I then stumbled on [llama.cpp's GBNF][2], which seems like an actively supported and developed method for accomplishing my goal. However, I assume GBNF is not exactly the same thing as Lark's flavor of EBNF. Would it be worthwhile to add some kind of export method that converts a Lark grammar into GBNF?

Finally, while looking through the Lark README I noticed [Outlines, which has its own support for generating structured output][3], this time built directly using Lark. This seems like the ideal solution.

Should I just use Outlines? Asking here in case I am missing something obvious, as it _is_ my first rodeo with LLMs.

[1]: https://github.com/r2d4/parserllm
[2]: https://github.com/ggml-org/llama.cpp/blob/master/grammars/README.md
[3]: https://dottxt-ai.github.io/outlines/latest/reference/generation/cfg/

**Comment by erezsh** [comment]
Hi @nchammas !

Personally I didn't use Outlines yet, but I've heard good things, so it's probably a good place to start.

**Reply by nchammas** [reply]
Just FYI that my initial experience with Outlines has been a [bit][1] [rough][2], and I see that [many others][3] are having regular problems with it specifically in the context of guiding generation using a grammar.

Other projects I have just come across in this space include [transformers-CFG][4] and [xgrammar][5], which both use GBNF grammars to guide generation. There is also [lm-format-enforcer][6], which offers a limited set of ways to guide generation, but there is an [open request to add full grammar support][7] in the GBNF format.

If I can't make progress with Outlines, I will experiment with GBNF and one of the other libraries that support it. It seems like GBNF is emerging as a de facto standard for grammars to guide LLM generation.

[1]: https://github.com/dottxt-ai/outlines/issues/1475
[2]: https://github.com/dottxt-ai/outlines/issues/1478
[3]: https://github.com/dottxt-ai/outlines/issues?q=is%3Aissue%20state%3Aopen%20grammar
[4]: https://github.com/epfl-dlab/transformers-CFG
[5]: https://github.com/mlc-ai/xgrammar
[6]: https://github.com/noamgat/lm-format-enforcer
[7]: https://github.com/noamgat/lm-format-enforcer/issues/78

**Reply by nchammas** [reply]
[This thread on huggingface/transformers][1] led me to many of the other projects in my reply just above.

[1]: https://github.com/huggingface/transformers/issues/25778

**Reply by nchammas** [reply]
OK, I just found out about [LLGuidance][1], which uses a [variant of Lark syntax][2] for its grammar to guide LLMs. And unlike Outlines, which requires a LALR(1) grammar, Guidance uses Earley.

Very promising, but my streak of bad luck [continues][3]. How are all of these projects so broken? Anyway, just sharing.

[1]: https://github.com/ggml-org/llama.cpp/blob/5e43f104cca1a14874e980326a506b44fde022b8/docs/llguidance.md
[2]: https://github.com/guidance-ai/llguidance/blob/9a9e6c68528ec485ec90a1b5bb4ffe7da49b508d/docs/syntax.md
[3]: https://github.com/guidance-ai/guidance/issues/1151

**Reply by erezsh** [reply]
> How are all of these projects so broken?

Probably all written with vibe-coding 😆 

But also everyone is "moving fast and breaking things", so maybe over time even good code can break, if it's not actively maintained.

Anyway, I'm glad you're documenting it here. It might be useful for others in the future. 👍 

I hope your luck improves!

**Reply by nchammas** [reply]
Just noting for the record that the [LMQL project][1] also has [an open issue about adding support for context-free grammars][2], with explicit mention of Lark as an option.

[1]: https://lmql.ai
[2]: https://github.com/eth-sri/lmql/issues/304

**Reply by erezsh** [reply]
Interesting. Maybe Lark should offer an API specifically to help next token predictions.

**Reply by nchammas** [reply]
Given what I've learned over the past week or so, I think Lark is well positioned to emerge as the de facto library for constraining LLM output via an EBNF grammar, perhaps replacing even GBNF.

If you have the interest and bandwidth, you may want to reach out to the maintainers of Outlines, Guidance, and LMQL to see if there is anything they need from Lark to enable better integrations. It's still early days in this space, and things are [evolving][2] [rapidly][1].

[1]: https://github.com/guidance-ai/guidance/issues/1151#issuecomment-2712125068
[2]: https://github.com/guidance-ai/guidance/pull/1150

**Reply by erezsh** [reply]
It looks to me like llguidance isn't actually using the Lark parser, but only using the Lark syntax for its grammars. And it's Rust, so it's unlikely they will call a Python library to constrain the tokens.

LMQL is more likely, but doesn't seem like they're prioritizing logit filtering / biasing.

It looks like Ollama are talking about adding an API for logits, though there's no clear schedule for it.

**Reply by nchammas** [reply]
Just a general FYI: Looks like supporting indentation-significant grammars is [particularly tricky for LLMs][1].

[1]: https://github.com/guidance-ai/llguidance/pull/128

**Reply by erezsh** [reply]
I spent some time thinking about how to generate indentation a few months ago. I don't know how helpful it would be, but I'll put it here:

The gist of the idea is to have multiple lexers, one for each indentation level, and use parser callbacks to move between these lexers. (similar to how the contextual-lexer in Lark moves between states)
That shifts the handling of indentation from column-counting into regular lexing (i.e. a state-machine), with only having to move between states.
While that might suggest a huge amount of lexers, in practice most real-world code has less than 16 levels of indentation. Furthermore, those lexers would all be identical to each other, except for the regexps of the INDENT and DEDENT terminals. For indentation, changing between lexers actually only requires changing the values of these 2 terminals.

(the explanation assumes it uses Lark, but maybe it can apply to other parsers as well)

**Reply by nchammas** [reply]
Interesting! This comment would be better appreciated by @mmoskal (the author of the LLGuidance PR I linked to just above).

(Apologies to Michał if the ping is not appropriate. Just thought you might find this interesting.)

**Reply by mmoskal** [reply]
Thanks @nchammas for bringing this up!

The lexer in LLGuidance uses a custom derivative-based regex [library](https://github.com/microsoft/derivre) where lexer automaton states are built dynamically from the set of currently allowed lexemes (terminals) (derived from Earley parser state). Thus, depending on the current indent level I can indicate that say lexeme `"\n  "` is NEWLINE, `"\n    "` is INDENT, and `"\n"` is DEDENT.

The problem is that multiple DEDENT lexemes can be generated from a single byte of input and I'm generally not set up for that (the price you pay for hyper-optimizing everything).

As for conversions, we have https://github.com/guidance-ai/llguidance/blob/main/python/llguidance/gbnf_to_lark.py and there is an inverse in vLLM https://github.com/vllm-project/vllm/blob/main/vllm/v1/structured_output/utils.py#L105

Note that gBNF will def. not handle indent/dedent (they don't even have `%ignore`; or for that matter a concept of a separate lexer). The Lark syntax is more expressive.

**Reply by nchammas** [reply]
Thank you for that explanation @mmoskal.

Just wanted to add another note here for @erezsh about yet another project I found in this space: [SynCode][1].

It looks like they too use Lark's syntax to support constraining LLM generation to an EBNF grammar, even claiming to handle indentation-sensitive output like Python well. They support [LALR(1) and LR(1) grammars][2].

[1]: https://github.com/structuredllm/syncode
[2]: https://github.com/structuredllm/syncode/tree/e2cf2c5261a63207718dfaab43b9d7d21ccc03b5/syncode/parsers/grammars#lalr1-or-lr1-grammars
